### PR TITLE
Disable overcommit accounting on E2 VMs (again)

### PR DESCRIPTION
**Pull request by hopkiw**

### CODE REVIEW

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 1]
```diff
@@ -65,17 +64,27 @@ def __init__(self, debug=False):
...
         host_key_types = self.instance_config.GetOptionString(
             'InstanceSetup', 'host_key_types')
         self._SetSshHostKeys(host_key_types=host_key_types)
+
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_boto_config'):
         self._SetupBotoConfig()
+
+      # machineType is e.g. u'projects/00000000000000/machineTypes/n1-standard-1'
+      machineType = self.metadata_dict['instance']['machineType'].split('/')[-1]
+      if machineType.startswith("e2-") and 'bsd' not in distro_name:  # Not yet supported on BSD.
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Line is greater than 80 characters.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 2]
```diff
@@ -65,17 +64,27 @@ def __init__(self, debug=False):
...
+
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_host_keys'):
         host_key_types = self.instance_config.GetOptionString(
             'InstanceSetup', 'host_key_types')
         self._SetSshHostKeys(host_key_types=host_key_types)
+
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_boto_config'):
         self._SetupBotoConfig()
+
+      # machineType is e.g. u'projects/00000000000000/machineTypes/n1-standard-1'
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Please end the comment with a period.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 3]
```diff
@@ -28,8 +28,7 @@
 from google_compute_engine import logger
 from google_compute_engine import metadata_watcher
 from google_compute_engine.boto import boto_config
-from google_compute_engine.compat import urlerror
-from google_compute_engine.compat import urlrequest
+from google_compute_engine.compat import distro_name, urlerror, urlrequest
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Per the style we've been following throughout this codebase, please put these on separate lines.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 4]
```diff
@@ -65,17 +64,27 @@ def __init__(self, debug=False):
...
             'InstanceSetup', 'host_key_types')
         self._SetSshHostKeys(host_key_types=host_key_types)
+
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_boto_config'):
         self._SetupBotoConfig()
+
+      # machineType is e.g. u'projects/00000000000000/machineTypes/n1-standard-1'
+      machineType = self.metadata_dict['instance']['machineType'].split('/')[-1]
+      if machineType.startswith("e2-") and 'bsd' not in distro_name:  # Not yet supported on BSD.
+        subprocess.call(["sysctl", "vm.overcommit_memory=1"])
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Please use single quotes instead of double quotes (everywhere).

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 5]
```diff
@@ -46,7 +46,7 @@ def testInstanceSetup(self, mock_logger, mock_watcher, mock_config):
     mock_logger_instance = mock.Mock()
     mock_logger.Logger.return_value = mock_logger_instance
     mock_watcher_instance = mock.Mock()
-    mock_watcher_instance.GetMetadata.return_value = {'hello': 'world'}
+    mock_watcher_instance.GetMetadata.return_value = {'hello': 'world', 'instance': {'machineType': 'fake'}}
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: Line length.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 6]
```diff
@@ -386,7 +386,7 @@ def testSetSshHostKeysFirstBoot(self, mock_listdir):
     ]
 
     self.assertEqual(sorted(mock_generate_key.mock_calls), expected_calls)
-    self.mock_instance_config.SetOption.assert_called_once_with(
+    self.mock_instance_config.SetOption.assert_called_with(
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: What's the reason for this change?

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py (hunk 2 above)
Comment (hopkiw) [reply packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: no, comments are not english sentences

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py (hunk 6 above)
Comment (hopkiw) [reply packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: idk

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 7]
```diff
@@ -65,17 +66,27 @@ def __init__(self, debug=False):
...
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_host_keys'):
         host_key_types = self.instance_config.GetOptionString(
             'InstanceSetup', 'host_key_types')
         self._SetSshHostKeys(host_key_types=host_key_types)
+
       if self.instance_config.GetOptionBool('InstanceSetup', 'set_boto_config'):
         self._SetupBotoConfig()
+
+      # machineType is e.g. 'projects/00000000000000/machineTypes/n1-standard-1'
+      machineType = self.metadata_dict['instance']['machineType'].split('/')[-1]
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: I think this belongs in a separate function. I don't see where we are currently testing that the machine type will be read correctly from metadata. Also, as written, I believe the current tests will be calling `subprocess.call(['sysctl', 'vm.overcommit_memory=1'])` since subprocess is not stubbed in many of the instance setup tests calling this.

Comment (illfelder) [reply packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Actually, because the fake metadata in the tests either doesn't have machine type, or has machine type with a value that does not start with `e2-` they won't be calling subprocess. Still doesn't seem ideal to me.

Comment (illfelder) [reply packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Something like this:
```
def _DisableOvercommit(self):
  """Disable overcommit accounting on E2 machine types."""

  # Expected machine type format:
  # 'projects/00000000000/machineTypes/n1-standard-1'.
  machine_type = self.metadata_dict['instance']['machineType'].split('/')[-1]
  if machine_type.startswith('e2-') and 'bsd' not in distro_name:
    subprocess.call(['sysctl', 'vm.overcommit_memory=1'])
```

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py [hunk 8]
```diff
@@ -65,22 +66,38 @@ def __init__(self, debug=False):
...
+
     if self.instance_config.GetOptionBool('InstanceSetup', 'set_multiqueue'):
       self._RunScript('google_set_multiqueue')
+
     try:
       self.instance_config.WriteConfig()
     except (IOError, OSError) as e:
       self.logger.warning(str(e))
 
+  def _DisableOvercommit(self):
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/instance_setup.py]: Please add a test that verifies subprocess is not called when the machine type is not E2 or distro name is bsd and a test to validate that subprocess is called when the conditions apply.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 9]
```diff
@@ -421,6 +428,27 @@ def testSetupBotoConfigLocked(self, mock_boto):
     instance_setup.InstanceSetup._SetupBotoConfig(self.mock_setup)
     self.mock_logger.warning.assert_called_once_with('Test Error')
 
+  @mock.patch('google_compute_engine.instance_setup.instance_setup.subprocess')
+  def testDisableOvercommitNonE2(self, mock_subprocess):
+    self.mock_setup.metadata_dict = {
+        'instance': {
+            'machineType': 'projects/00000000000/machineTypes/n1-standard-1'
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: Please add a comma to the end of the line since it is an entry in the dictionary. (Also apply to the test case below).

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 10]
```diff
@@ -421,6 +428,27 @@ def testSetupBotoConfigLocked(self, mock_boto):
...
+
+  @mock.patch('google_compute_engine.instance_setup.instance_setup.subprocess')
+  def testDisableOvercommitE2(self, mock_subprocess):
+    self.mock_setup.metadata_dict = {
+        'instance': {
+            'machineType': 'projects/00000000000/machineTypes/e2-standard-1'
+        }
+    }
+    instance_setup.InstanceSetup._DisableOvercommit(self.mock_setup)
+    mock_subprocess.call.assert_called_once_with(['sysctl',
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: Please line wrap after the open paren.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 11]
```diff
@@ -421,6 +428,38 @@ def testSetupBotoConfigLocked(self, mock_boto):
...
+    )
+
+  @mock.patch('google_compute_engine.instance_setup.instance_setup.subprocess')
+  def testDisableOvercommitBSD(self, mock_subprocess):
+    self.mock_setup.metadata_dict = {
+        'instance': {
+            'machineType': 'projects/00000000000/machineTypes/e2-standard-1',
+        }
+    }
+    instance_setup.InstanceSetup._DisableOvercommit(self.mock_setup, "bsd")
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: Single quote please.

File: packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py [hunk 12]
```diff
@@ -421,6 +428,38 @@ def testSetupBotoConfigLocked(self, mock_boto):
...
+  def testDisableOvercommitE2(self, mock_subprocess):
+    self.mock_setup.metadata_dict = {
+        'instance': {
+            'machineType': 'projects/00000000000/machineTypes/e2-standard-1',
+        }
+    }
+    instance_setup.InstanceSetup._DisableOvercommit(self.mock_setup)
+    mock_subprocess.call.assert_called_once_with(
+        ['sysctl', 'vm.overcommit_memory=1']
+    )
```
Comment (illfelder) [inline packages/python-google-compute-engine/google_compute_engine/instance_setup/tests/instance_setup_test.py]: Please move the close brace onto the line above.
